### PR TITLE
Rearrange SyncWorker code & Fix bugs

### PR DESCRIPTION
**Pull request by michel-tricot**
## What
There are too many abstractions and factories when it it could actually just be first class Tap & Target objects.
Fixes the following bugs:
- invalid schema validation on JsonNode
- invalid closing of stream
- invalid process shutdown
- error when displaying logs on error
- properly notify that tap is finished to the target
- invalid detection of end of tap stream
- invalid propagation of process termination status

## Recommended reading order
1. `SingerSyncWorker.java`
1. `SingerTap.java`
1. `SingerTarget.java`
1. the rest

### CODE REVIEW

File: dataline-workers/src/main/java/io/dataline/workers/process/DockerProcessBuilderFactory.java [hunk 1]
```diff
@@ -63,7 +64,7 @@ public ProcessBuilder create(final Path jobRoot, final String imageName, final S
             imageName);
     cmd.addAll(Arrays.asList(args));
 
-    LOGGER.debug("Preparing command: {}", cmd);
+    LOGGER.debug("Preparing command: {}", Joiner.on(" ").join(cmd));
```
Comment (michel-tricot) [inline dataline-workers/src/main/java/io/dataline/workers/process/DockerProcessBuilderFactory.java:67]: Changed the logging so it is possible to copy & paste the the command in a console

File: dataline-workers/src/main/java/io/dataline/workers/singer/SingerJsonStreamFactory.java [hunk 2]
```diff
@@ -59,17 +58,13 @@ public SingerJsonStreamFactory() {
...
-  private SingerMessage parseJsonOrNull(String record) {
-    Optional<SingerMessage> message = Jsons.tryDeserialize(record, SingerMessage.class);
-    if (message.isPresent()) {
-      return message.get();
-    } else {
-      LOGGER.info("Record was not a json representation of a SingerMessage. Record: {}", record);
-      return null;
-    }
+    return bufferedReader
+        .lines()
```
Comment (michel-tricot) [inline dataline-workers/src/main/java/io/dataline/workers/singer/SingerJsonStreamFactory.java:62]: this is one of the only place where the behavior might change.

I changed it because the previous code was deserializing twice and the tryDeserialize was actually useless since the validator was doing an unguarded deserialization.

File: dataline-workers/src/main/java/io/dataline/workers/singer/SingerSyncWorker.java [hunk 3]
```diff
@@ -22,9 +22,8 @@
  * SOFTWARE.
  */
 
-package io.dataline.workers;
+package io.dataline.workers.singer;
```
Comment (michel-tricot) [inline dataline-workers/src/main/java/io/dataline/workers/singer/SingerSyncWorker.java:25]: Renamed it to SingerSyncWorker because it can only work with singer messages

File: dataline-workers/src/main/java/io/dataline/workers/singer/SingerSyncWorker.java [hunk 4]
```diff
@@ -68,16 +70,23 @@ public DefaultSyncWorker(TapFactory<SingerMessage> singerTapFactory,
 
     final SingerMessageTracker singerMessageTracker = new SingerMessageTracker();
 
-    try (Stream<SingerMessage> tap = singerTapFactory.create(tapConfig, jobRoot);
```
Comment (michel-tricot) [inline dataline-workers/src/main/java/io/dataline/workers/singer/SingerSyncWorker.java:71]: This code doesn't work today

File: dataline-workers/src/main/java/io/dataline/workers/singer/SingerSyncWorker.java [hunk 5]
```diff
@@ -68,16 +70,23 @@ public DefaultSyncWorker(TapFactory<SingerMessage> singerTapFactory,
...
 
+      singerTap.stop();
+      singerTarget.stop();
     } catch (Exception e) {
-      LOGGER.debug(
-          "Sync worker failed. Tap error log: {}.\n Target error log: {}",
-          IOs.readFile(jobRoot, TAP_ERR_LOG),
-          IOs.readFile(jobRoot, TARGET_ERR_LOG));
+      LOGGER.error("Sync worker failed. Tap error log: {}.\n Target error log: {}",
+          Files.exists(jobRoot.resolve(TAP_ERR_LOG)) ? IOs.readFile(jobRoot, TAP_ERR_LOG) : "<null>",
```
Comment (michel-tricot) [inline dataline-workers/src/main/java/io/dataline/workers/singer/SingerSyncWorker.java:87]: Exception where not displayed because they happened before the files were created

File: dataline-workers/src/main/java/io/dataline/workers/singer/SingerTarget.java [hunk 6]
```diff
@@ -27,56 +27,72 @@
...
-      // the TargetConsumer is responsible for closing this.
-      BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(targetProcess.getOutputStream(), Charsets.UTF_8));
-      return new TargetConsumer(writer, targetProcess);
+      writer = new BufferedWriter(new OutputStreamWriter(targetProcess.getOutputStream(), Charsets.UTF_8));
     } catch (Exception e) {
       throw new RuntimeException(e);
     }
   }
 
+  public void consume(SingerMessage message) throws IOException {
```
Comment (michel-tricot) [inline dataline-workers/src/main/java/io/dataline/workers/singer/SingerTarget.java:83]: This is the code from the TargetConsumer

File: dataline-workers/src/main/java/io/dataline/workers/singer/SingerSyncWorker.java (hunk 3 above)
Comment (cgardens) [reply dataline-workers/src/main/java/io/dataline/workers/singer/SingerSyncWorker.java:25]: true given the changes you made to the constructor, but why are you changing the implementation for that to be the case? What in this class makes it important to pass in a SingerTap and a SingerTarget as opposed to just a Tap and Target?

File: dataline-workers/src/main/java/io/dataline/workers/singer/SingerSyncWorker.java (hunk 4 above)
Comment (cgardens) [reply dataline-workers/src/main/java/io/dataline/workers/singer/SingerSyncWorker.java:71]: what's the issue?

File: dataline-workers/src/main/java/io/dataline/workers/singer/SingerSyncWorker.java [hunk 7]
```diff
@@ -68,16 +70,23 @@ public DefaultSyncWorker(TapFactory<SingerMessage> singerTapFactory,
 
     final SingerMessageTracker singerMessageTracker = new SingerMessageTracker();
 
-    try (Stream<SingerMessage> tap = singerTapFactory.create(tapConfig, jobRoot);
-        CloseableConsumer<SingerMessage> consumer = singerTargetFactory.create(targetConfig, jobRoot)) {
+    try {
+      singerTarget.start(targetConfig, jobRoot);
```
Comment (cgardens) [inline dataline-workers/src/main/java/io/dataline/workers/singer/SingerSyncWorker.java:74]: this is pushing a bunch of problems that were caught by the compiler into things that can only be caught at runtime. specifically it's adding more statefulness to the object with this start concept why do we prefer this?

File: dataline-workers/src/main/java/io/dataline/workers/singer/SingerTap.java [hunk 8]
```diff
@@ -115,48 +112,42 @@ public SingerTapFactory(final String imageName,
...
+    tapProcess =
+        pbf.create(jobRoot, imageName, cmd)
+            .redirectError(jobRoot.resolve(SingerSyncWorker.TAP_ERR_LOG).toFile())
+            .start();
 
     bufferedReader = new BufferedReader(new InputStreamReader(tapProcess.getInputStream()));
+    messageIterator = streamFactory.create(new BufferedReader(new InputStreamReader(tapProcess.getInputStream()))).iterator();
+  }
+
+  public boolean hasNext() {
```
Comment (cgardens) [inline dataline-workers/src/main/java/io/dataline/workers/singer/SingerTap.java:124]: if this thing is an iterator and is used as an iterator why aren't we taking advantage of that interface in java and at least calling it that?

Comment (cgardens) [reply dataline-workers/src/main/java/io/dataline/workers/singer/SingerTap.java:124]: also why are you moving from stream to iterator? i thought we had agreed that either was fine. what is changing your mind?

File: dataline-workers/src/main/java/io/dataline/workers/singer/SingerTarget.java (hunk 6 above)
Comment (cgardens) [reply dataline-workers/src/main/java/io/dataline/workers/singer/SingerTarget.java:83]: same. if this is a consumer, why aren't we using that interface?

**Review by cgardens — CHANGES_REQUESTED** [review]
I'm not comfortable with the new Taps and Targets. They're trading trading off less code for more errors that can only be caught at run time and objects with more complicated lifecyles. I think the complicated object lifecyles are going to bite us.

Definitely happy to see this code simplified, but I don't like this specific trade off.

Let's discuss offline when you're ready. 🙂

**Review by jrhizor — COMMENTED** [review]
Generally a 👍 (removing unnecessary layers is nice) but I share Charles's concerns.

File: dataline-workers/src/main/java/io/dataline/workers/singer/SingerTap.java (hunk 8 above)
Comment (cgardens) [reply dataline-workers/src/main/java/io/dataline/workers/singer/SingerTap.java:124]: discussed offline: 
* will implement interator interface
* empirically it was hard to debug this as a stream as iterators are more debug friendly in java, hence the switch to iterator.

File: dataline-workers/src/main/java/io/dataline/workers/singer/SingerSyncWorker.java (hunk 7 above)
Comment (cgardens) [reply dataline-workers/src/main/java/io/dataline/workers/singer/SingerSyncWorker.java:74]: discussed offline: there were issues around when streams are being closed, and so at this early stage empirically it was easier to debug and make sure our behavior was a correct by explicitly handling the lifecycle in one class without try-resources, etc.

**Review by sherifnada — CHANGES_REQUESTED** [review]
I agree with Charles' points here. Also, the code being removed is pretty minimal/simple, so the upside isn't very clear.

File: dataline-workers/src/main/java/io/dataline/workers/singer/SingerSyncWorker.java (hunk 3 above)
Comment (cgardens) [reply dataline-workers/src/main/java/io/dataline/workers/singer/SingerSyncWorker.java:25]: discussed offline: the DefaultSyncWorker is Singer protocol specific but not specific to a singer tap and target. This will get renamed to something like SingerProtocolSyncWorker and will take in SingerProtocolTap and SingerProtocolTarget.

**Comment by michel-tricot** [comment]
I discussed with charles about it. The main reason for removing these layers it is very hard to follow the lifecycle of processes, and streams. 

Here is an updated version. 

This version fixes several bugs. see pr descriptions for more details

File: dataline-workers/src/main/java/io/dataline/workers/protocols/singer/SingerTap.java [hunk 9]
```diff
@@ -22,15 +22,24 @@
...
 import io.dataline.config.StandardTapConfig;
+import io.dataline.singer.SingerMessage;
+import io.dataline.workers.InvalidCredentialsException;
+import java.io.IOException;
 import java.nio.file.Path;
-import java.util.stream.Stream;
+import java.util.Optional;
 
-public interface TapFactory<T> {
+public interface SingerTap extends AutoCloseable {
```
Comment (cgardens) [inline dataline-workers/src/main/java/io/dataline/workers/protocols/singer/SingerTap.java:34]: I think I preferred SingerProtocol tap just to avoid confusion. but nbd if it's too verbose.

File: dataline-workers/src/main/java/io/dataline/workers/protocols/singer/DefaultSingerTarget.java [hunk 10]
```diff
@@ -22,61 +22,100 @@
...
-      return new TargetConsumer(writer, targetProcess);
+      writer = new BufferedWriter(new OutputStreamWriter(targetProcess.getOutputStream(), Charsets.UTF_8));
     } catch (Exception e) {
+      // TODO: we should probably do some cleanup here.
       throw new RuntimeException(e);
     }
   }
 
+  @Override
+  public void accept(SingerMessage message) throws IOException {
```
Comment (cgardens) [inline dataline-workers/src/main/java/io/dataline/workers/protocols/singer/DefaultSingerTarget.java:88]: `    Preconditions.checkState(!endOfStream);`?

**Comment by michel-tricot** [comment]
I wan't to add more testing on Tap/Target/Worker. I will do it in a following PR as these issues are show blocker.

File: dataline-workers/src/main/java/io/dataline/workers/protocols/singer/SingerSyncWorker.java [hunk 11]
```diff
@@ -68,16 +71,27 @@ public DefaultSyncWorker(TapFactory<SingerMessage> singerTapFactory,
...
 
-    try (Stream<SingerMessage> tap = singerTapFactory.create(tapConfig, jobRoot);
-        CloseableConsumer<SingerMessage> consumer = singerTargetFactory.create(targetConfig, jobRoot)) {
+    try (singerTarget; singerTap) {
 
-      tap.takeWhile(record -> !cancelled.get()).peek(singerMessageTracker).forEach(consumer);
+      singerTarget.start(targetConfig, jobRoot);
+      singerTap.start(tapConfig, jobRoot);
+
+      while (!cancelled.get() && singerTap.isFinished()) {
```
Comment (cgardens) [inline dataline-workers/src/main/java/io/dataline/workers/protocols/singer/SingerSyncWorker.java:79]: should this be? `      while (!cancelled.get() && !singerTap.isFinished()) {`

**Review by cgardens — APPROVED** [review]
lgtm.

File: dataline-workers/src/main/java/io/dataline/workers/protocols/singer/SingerSyncWorker.java (hunk 11 above)
Comment (michel-tricot) [reply dataline-workers/src/main/java/io/dataline/workers/protocols/singer/SingerSyncWorker.java:79]: :) damn :)

Comment (michel-tricot) [reply dataline-workers/src/main/java/io/dataline/workers/protocols/singer/SingerSyncWorker.java:79]: I renamed it last minute it was a negative before

**Review by jrhizor — APPROVED** [review]
